### PR TITLE
Unification algorithm modification (projections vs canonical structures)

### PR DESCRIPTION
**Pull request by Tragicus**
Draft for a modification of the unification algorithm, to see how much it breaks the CI. For now, it catches unification problems where one of the term is a projection applied to a constructor (modulo weak head reduction) and simplifies these terms.

Fixes / closes #????

- [ ] Added / updated **test-suite**.

- [ ] Added **changelog**.
- [ ] Added / updated **documentation**.
  
  - [ ] Documented any new / changed **user messages**.
  - [ ] Updated **documented syntax** by running `make doc_gram_rsts`.

- [ ] Opened **overlay** pull requests.

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

File: pretyping/evarconv.ml [hunk 1]
```diff
@@ -568,6 +568,38 @@ let infer_conv_noticing_evars ~pb ~ts env sigma t1 t2 =
     if !has_evar then None
     else Some (UnifFailure (sigma, UnifUnivInconsistency e))
 
+let pr_econstr = ref (fun _ _ _ -> Pp.str "unable to print econstr")
```
Comment (SkySkimmer) [inline pretyping/evarconv.ml:571]: you don't need this, there's already the same thing at Termops.Internal

**Comment by SkySkimmer** [comment]
What problem does this solve?

File: pretyping/evarconv.ml (hunk 1 above)
Comment (Tragicus) [reply pretyping/evarconv.ml:571]: Excellent, thank you.

**Comment by Tragicus** [comment]
Thank you for looking into it.

With the current algorithm, unification of `proj _` with something always (afaikt) uses canonical structures, which in my usecase introduces unification variables, which causes "a few" problems. Without going into too much details, I had one particular case where one of these unification variables found its way into both terms to unify, which made the algorithm fail.

Besides, the debug trace was unreadable because the algorithm spent all its time instantiating things using canonical structures and checking that the solution was correct, so I am hopeful that this will improve performances, when it works.

**Comment by SkySkimmer** [comment]
There are failures in the test suite, do you want to see the full CI anyway?

**Comment by SkySkimmer** [comment]
I guess the change is that in 
~~~coq

Set Primitive Projections.
Record R := { foo : nat; bar : nat }.

Canonical zeros : R := {| foo := 0; bar := 0 |}.

Goal exists x, foo x = 0.
Proof.
  eexists {| foo := _; bar := _ |}.
  exact eq_refl.
Qed.
~~~
it will solve `hole for foo == 0` instead of solving `{| foo := _; bar := _ |} == zeros`, and the hole for bar will be left unsolved?

**Comment by Tragicus** [comment]
> There are failures in the test suite, do you want to see the full CI anyway?

No, I forgot that I had access to the test suite locally, I will push again when it works.

**Comment by SkySkimmer** [comment]
You can also run the CI locally, although running everything probably takes more time than it's worth.
For primitive projections HoTT is a major user and not too big, so if you want you can do `make ci-hott` (after make world, and if you change something after make ci-hott you need to `cd _build_ci/hott && make clean` before doing make ci-hott again)

**Comment by Tragicus** [comment]
> I guess the change is that in
> 
> ```coq
> Set Primitive Projections.
> Record R := { foo : nat; bar : nat }.
> 
> Canonical zeros : R := {| foo := 0; bar := 0 |}.
> 
> Goal exists x, foo x = 0.
> Proof.
>   eexists {| foo := _; bar := _ |}.
>   exact eq_refl.
> Qed.
> ```
> 
> it will solve `hole for foo == 0` instead of solving `{| foo := _; bar := _ |} == zeros`, and the hole for bar will be left unsolved?

This is what I expect to see. In this case, it produces a shelved goal, which I would prefer to avoid, but I remember seeing terms with holes that I definitely wanted to reduce, so it will require some thought (which is why I wanted to look at the CI).

**Comment by CohenCyril** [comment]
@coqbot bench

**Pull request closed without merging**